Add support for Palantir Java Formatter

eirnym opened the pull request:
[Palantir Java Formatter](https://github.com/palantir/palantir-java-format) has a bit different settings than Google Java Format.

As the original project doesn't provide standalone version, [repackager CI](https://github.com/jsonschema2dataclass/palantir-cli) has been used.

eirnym commented on language_formatters_pre_commit_hooks/pretty_format_java.py:
```diff
@@ -128,7 +185,52 @@ def pretty_format_java(argv: typing.Optional[typing.List[str]] = None) -> int:
+
+    cmd_args = [
+        "java",
+        "-jar",
```
Palantir Java Formatter has additional permissions inside manifest and they aren't required to be provided separately

eirnym commented on tests/pretty_format_java_test.py:
```diff
@@ -52,22 +65,58 @@ def test__download_google_java_formatter_jar(ensure_download_possible, version):
 
 @pytest.mark.skipif(condition=get_jdk_version() < Version("16"), reason="Skipping test because it requires Java JDK 16+")
-def test_pretty_format_java_up_to_1_9_is_not_allowed_on_jdk_16_and_above(undecorate_method):
+def test_google_pretty_format_java_up_to_1_9_is_not_allowed_on_jdk_16_and_above(undecorate_method):
```
added `google` and `palantir` to distinguish tests.

macisamuele reviewed (approved):
Change looks good and thanks for updating the automation to automatically update the tool versions.
If tests are green I'll be happy to merge and release this change

eirnym commented:
Palantir Java formatter currently requires Java Version < 21 and is not available if java version is above 21, so coverage won't show 100% if it's the case.

macisamuele commented:
If this is the case we need to make sure that
1) at runtime we check the jdk in use and bail out if not on the proper version (there is already an helper for that)
2) have unit test verifying that
3) have tests that run the tool end to end with pytest.skipif tk make sure that the test is only run if new sdk available

eirnym commented:
I think I have this test, I'll double check it.

eirnym commented:
Anyway this won't help with coverage that you'd like to be 100%

eirnym commented:
Do you mind if I'll split `google formatter` and `palantir`?

macisamuele commented:
> Do you mind if I'll split tests for `google formatter` and `palantir` into separate files?

I would rather prefer to keep the tests of a single module into a single test file, it makes it managing them more predictable.

Test coverage can be kept high, it is sufficient to check what the test is mentioning as not covered lines
```
  py310: commands[2]> coverage report
  Name                                                         Stmts   Miss  Cover   Missing
  ------------------------------------------------------------------------------------------
  language_formatters_pre_commit_hooks/pretty_format_java.py      54      1    98%   220
  ------------------------------------------------------------------------------------------
  TOTAL                                                          356      1    99%
```

Checking the code I see that the following line is not covered
https://github.com/macisamuele/language-formatters-pre-commit-hooks/blob/881f4e1e7ed0641225d1a5c9267da927d3fe97f7/language_formatters_pre_commit_hooks/pretty_format_java.py#L220

The coverage missing is due to the fact that we cannot guarantee that end-to-end test is run with Java 21 in all the environments, hence let's add `# pragma: nocover` for now to silence the coverage report and we can check it on a second round

eirnym commented:
@macisamuele I've done as you said.

* rebased to master
* added nocover for replace flag
* Fixed one test name
* passed JAVA_HOME to tox (as on a local computer many Java version are installed)

For java pre 21, coverage is 100%, while on java 21 palantir routine won't pass because of test precondition.

eirnym commented:
@macisamuele any news?

macisamuele commented on language_formatters_pre_commit_hooks/pretty_format_java.py:
```diff
@@ -130,6 +187,51 @@ def pretty_format_java(argv: typing.Optional[typing.List[str]] = None) -> int:
+    # Palantir Java formatter is not compatible with Java 21 and above yet.
+    # https://github.com/palantir/palantir-java-format/issues/977 contains
+    # more links for Java compatibility
+    assert_max_jdk_version(Version("21.0"))  # pragma: no cover
```
nit: If you have any test that would try this logic this line will be covered. No reason to exclude it for coverage

```suggestion
    assert_max_jdk_version(Version("21.0"))
```

eirnym replied:
I've covered this line, but lines below I can't cover on Java 21. Do you have any suggestions?

macisamuele reviewed (approved):
Change looks overall good.

eirnym commented on setup.cfg:
```diff
@@ -41,5 +41,6 @@ exclude =
 [options.package_data]
 language_formatters_pre_commit_hooks =
     google_java_formatter.version
+    palantir_java_formatter.version
```
I'll rename it. My thought was to rename Google Java formatter.

mxr commented on .pre-commit-hooks.yaml:
```diff
@@ -13,12 +13,12 @@
   language: python
   types: [java]
-  # this is needed because the hook downloads google-java-formatter and
+  # this is needed because the hook downloads palanrir/google-java-formatter and
```
```suggestion
  # this is needed because the hook downloads palantir/google-java-formatter and
```

mxr replied:
I see this is resolved but the typo is still there - friendly ping :)

mxr commented on tests/pretty_format_java_test.py:
```diff
@@ -43,6 +44,18 @@ def test__download_google_java_formatter_jar(ensure_download_possible, version):
+)
+@pytest.mark.integration
+def test__download_palantir_java_formatter_jar(ensure_download_possible, version):  # noqa: F811
+    # Test that we can download different version of the Google Java Formatter
```
```suggestion
    # Test that we can download different version of the Palantir Java Formatter
```

CopperGiraffe commented:
Hello there, Palantir support would be great, is there any news on the subject?

eirnym commented:
@macisamuele Palantir fixed the issue and now supports Java 21. I updated version, rebased and added tests. Could you please review it again?

mxr commented on .github/workflows/bump_external_releases.py:
```diff
@@ -68,6 +68,13 @@ def bump_google_java_formatter():
+
 if __name__ == "__main__":
     something_is_bumped = False
     for bumper in (bump_ktfmt, bump_ktlint, bump_google_java_formatter):
```
You are missing `bump_palantir_java_formatter` here. Instead of continuing to update this line, you could automatically call all bumping functions as follows:

```py
something_is_bumped = False
for name, obj in tuple(locals().items()):
    if callable(obj) and name.startswith("bump") and name != "bump_release":
        something_is_bumped |= obj()
```

eirnym replied:
done

eirnym replied:
I'll follow Zen Of Python (_"Explicit is better than implicit"_)

I'd leave this refactoring to other PR, as I personally don't agree with querying locals. it may lead to problems which hard to test.

mxr replied:
No problem!

mxr commented on language_formatters_pre_commit_hooks/pretty_format_java.py:
```diff
@@ -50,22 +51,71 @@ def get_urls(_version: str) -> typing.List[str]:
 
 
+def _download_palantir_java_formatter_jar(version: str) -> str:  # pragma: no cover
+    def get_urls(_version: str) -> typing.List[str]:
+        # Links extracted from https://github.com/jsonschema2dataclass/palantir-cli
+        return [
+            "https://github.com/jsonschema2dataclass/palantir-cli/releases/download/"
+            "{version}/palantir-cli-{version}-standalone.jar".format(
+                version=_version,
+            ),
+        ]
+
+    possible_urls = get_urls(version)
```
I know this is copied from other parts of the code but IMO I would inline this function

macisamuele commented:
Please address coverage and then ready to merge.
```
language_formatters_pre_commit_hooks/pretty_format_java.py      59      3    95%   233-239
```

eirnym commented:
@macisamuele Finally nailed it!

eirnym commented:
Any news?

macisamuele commented:
Had a very busy time in the last days.
Enabled tests to run and if they pass this PR will be merged.
Tomorrow I'll check the status of this and release a new version if the change was merged

macisamuele commented:
Change is merged , at the same time pre-commit were not really fixed.
Going to fix them directly on master and take care of the release

eirnym commented:
Thank you!
